### PR TITLE
Fix access token attributes for federated user

### PR DESCRIPTION
**Pull request by shashimalcse**
### Proposed changes in this pull request

Previously JWTAccessTokenAttributesOIDCClaimHandler resolve attributes from the userstore level and JWT token will not contain Access Token attributes in the federation login flow. This PR improve the JWTAccessTokenAttributesOIDCClaimHandler to handle attributes of the federated user. Here we have move common methods from JWTAccessTokenAttributesOIDCClaimHandler and DefaultOIDCClaimCallbackHandler to OIDCClaimUtil.

### Related Issues
- https://github.com/wso2/product-is/issues/22009

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/openidconnect/JWTAccessTokenOIDCClaimsHandlerTest.java [hunk 1]
```diff
@@ -236,6 +236,28 @@ public void testHandleCustomClaimsWithoutJWTAccessTokenClaimsForOAuthTokenReqMsg
...
+    public void testHandleCustomClaimsWithoutJWTAccessTokenClaimsForOAuthAuthzReqMsgContext() throws Exception {
+
+        try (MockedStatic<JDBCPersistenceManager> jdbcPersistenceManager = mockStatic(JDBCPersistenceManager.class);
+             MockedStatic<OAuthServerConfiguration> oAuthServerConfiguration = mockStatic(
+                     OAuthServerConfiguration.class)) {
+            OAuthServerConfiguration oauthServerConfigurationMock = mock(OAuthServerConfiguration.class);
+            oAuthServerConfiguration.when(OAuthServerConfiguration::getInstance)
+                    .thenReturn(oauthServerConfigurationMock);
+            try (MockedStatic<OAuth2Util> oAuth2Util = mockStatic(OAuth2Util.class)) {
+                oAuth2Util.when(() -> OAuth2Util.getAppInformationByClientId(any(), any())).thenReturn(
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/openidconnect/JWTAccessTokenOIDCClaimsHandlerTest.java:249]: Let's change any to any(RequiredClass.class)

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/OIDCClaimUtil.java [hunk 2]
```diff
@@ -692,4 +715,531 @@ private static Map<String, Object> getUserClaimsInOidcDialect(Map<String, String
...
+     *
+     * @param claimKey Claim key.
+     * @param claimValue Claim value.
+     * @param multiAttributeSeparator Multi attribute separator.
+     * @return True if claim value is multi valued, false otherwise.
+     */
+    private static boolean isMultiValuedAttribute(String claimKey, String claimValue, String multiAttributeSeparator) {
+
+        // Address claim contains multi attribute separator but its not a multi valued attribute.
+        if (claimKey.equals(ADDRESS)) {
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/OIDCClaimUtil.java:1235]:
```suggestion
        if (ADDRESS.equals(claimKey)) {
```

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/OIDCClaimUtil.java [hunk 3]
```diff
@@ -692,4 +715,531 @@ private static Map<String, Object> getUserClaimsInOidcDialect(Map<String, String
...
+    }
+
+    /**
+     * Get the tenant domain of the service provider.
+     *
+     * @param requestMsgCtx OAuthAuthzReqMessageContext.
+     * @return Tenant domain of the service provider.
+     */
+    public static String getServiceProviderTenantDomain(OAuthAuthzReqMessageContext requestMsgCtx) {
+        String spTenantDomain = (String) requestMsgCtx.getProperty(MultitenantConstants.TENANT_DOMAIN);
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/OIDCClaimUtil.java:1162]: add line after method signature.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/JWTAccessTokenOIDCClaimsHandler.java [hunk 4]
```diff
@@ -91,75 +84,195 @@ public JWTClaimsSet handleCustomClaims(JWTClaimsSet.Builder builder, OAuthAuthzR
...
+     *
+     * @param requestMsgCtx Token request message context
+     * @return Mapped claimed
+     * @throws OAuthSystemException
+     */
+    private Map<String, Object> getUserClaimsInOIDCDialect(OAuthTokenReqMessageContext requestMsgCtx)
+            throws OAuthSystemException, IdentityOAuth2Exception {
+        // Map<"email", "peter@example.com">
+        Map<String, Object> userClaimsInOIDCDialect;
+        // Get any user attributes that were cached against the access token
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/JWTAccessTokenOIDCClaimsHandler.java:102]:
```suggestion
        // Get all user attributes that were cached against the access token
```
WDYT?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/OIDCClaimUtil.java [hunk 5]
```diff
@@ -692,4 +715,531 @@ private static Map<String, Object> getUserClaimsInOidcDialect(Map<String, String
...
+    public static Map<ClaimMapping, String> getCachedUserAttributes(OAuthTokenReqMessageContext requestMsgCtx)
+            throws OAuthSystemException, IdentityOAuth2Exception {
+
+        Map<ClaimMapping, String> userAttributes = getUserAttributesCachedAgainstAuthorizationCode(
+                getAuthorizationCode(requestMsgCtx));
+        if (log.isDebugEnabled()) {
+            log.debug("Retrieving claims cached against authorization_code for user: " +
+                    requestMsgCtx.getAuthorizedUser());
+        }
+        if (isEmpty(userAttributes)) {
```
Comment (Thumimku) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/OIDCClaimUtil.java:737]: line 749 also checking same condition right? this is getting repeated in this method please check

Comment (shashimalcse) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/OIDCClaimUtil.java:737]: here if userAttributes are empty we trying to get userAttributes 742, then after we check empty

**Pull request closed without merging**